rust/feature-flags/src/flag_matching.rs

```diff
@@ -34,42 +179,48 @@ pub struct Person {
 // and then we can easily refactor stuff around.
 // #[derive(Debug)]
 pub struct FeatureFlagMatcher {
-    // pub flags: Vec<FeatureFlag>,
     pub distinct_id: String,
-    pub database_client: Option<Arc<dyn DatabaseClient + Send + Sync>>,
-    // TODO do I need cached_properties, or do I get them from the request?
-    // like, in python I get them from the request.  Hmm.  Let me try that.
-    // OH, or is this the FlagMatcherCache.  Yeah, so this is the flag matcher cache
-    cached_properties: Option<HashMap<String, Value>>,
+    pub team_id: TeamId,
+    pub database_client: Option<DatabaseClientArc>,
+    // TODO should all the properties by cached in the same cache?
+    // Like what's the point of having separate caches for person and group properties,
+    // especially considering that they don't appear to persist across multiple instances of the matcher?
+    cached_person_properties: Option<HashMap<String, Value>>,
+    cached_group_properties: Option<HashMap<String, Value>>,
     person_property_overrides: Option<HashMap<String, Value>>,
-    // TODO handle group properties
-    // group_property_overrides: Option<HashMap<String, HashMap<String, Value>>>,
+    group_property_overrides: Option<HashMap<String, HashMap<String, Value>>>,
+    flag_matcher_cache: Arc<FlagsMatcherCache>, // TODO rename to something more descriptive – this is the cache for group type mappings
 }
 
 const LONG_SCALE: u64 = 0xfffffffffffffff;
 
 impl FeatureFlagMatcher {
     pub fn new(
         distinct_id: String,
-        database_client: Option<Arc<dyn DatabaseClient + Send + Sync>>,
+        team_id: TeamId,
+        database_client: Option<DatabaseClientArc>,
         person_property_overrides: Option<HashMap<String, Value>>,
-        // group_property_overrides: Option<HashMap<String, HashMap<String, Value>>>,
+        group_property_overrides: Option<HashMap<String, HashMap<String, Value>>>,
+        cache: Option<Arc<FlagsMatcherCache>>,
     ) -> Self {
         FeatureFlagMatcher {
-            // flags,
             distinct_id,
-            database_client,
-            cached_properties: None,
+            team_id,
+            database_client: database_client.clone(),
+            cached_person_properties: None,
+            cached_group_properties: None,
             person_property_overrides,
-            // group_property_overrides,
+            group_property_overrides,
+            flag_matcher_cache: cache
+                .unwrap_or_else(|| Arc::new(FlagsMatcherCache::new(team_id, database_client))),
```

Comment:
that shouldn't happen, in django it's created once per request -> https://github.com/PostHog/posthog/blob/master/posthog/models/feature_flag/flag_matching.py#L730 (not flagMatcher instance). I agree its still not great (but has been minimal impact perf wise) - this isn't a big cache. The weak goal was not to fetch multiple times per request (each flag might need this)

I think eventual goal was to get these group types in redis as well, and then the FlagsMatcherCache just gets it from redis more often than not. Never got there since it wasn't the most critical part to fix.

Reply:
> the weak goal was not to fetch multiple times per request (each flag might need this)

Ah, I think this unlocks why I felt confused, but that totally makes sense.  Since each request can evaluate `n` flags, it makes sense to establish a cache of group mappings right out of the gate so that we can reference it throughout the flag evaluation process.